### DISCUSSION TITLE
mutate themeConfig in plugin onInitialized hook

### DISCUSSION THREAD

**Discussion by superewald** [vuepress/core#705] [Q&A] [answered]
Hey there,
I want to create a plugin which creates the `themeConfig.navbar` config automaticly based on file structure and frontmatter data (similiar to [vuepress-plugin-autonav](https://github.com/webmasterish/vuepress-plugin-autonav), just for VuePress v2).

I created a simple test plugin like this:
```js
const {path} = require('@vuepress/utils')

module.exports = (config, ctx) => {
    return {
        name: 'autonav',
        onInitialized(app) {
            // double assignment is just to make sure im not using wrong context
            app.options.themeConfig.navbar = ctx.options.themeConfig.navbar = [
                 { text: 'hello', link: '/'}
            ]
        }
    }
}
```

The plugin is loading but changing the configuration has no effect on the navbar. Where is the right place to mutate the navbar config from within a plugin ?
Thanks in advance!

**Comment by Mister-Hope** [comment, marked as answer]
See @vuepress/plugin-theme-data, that's what are using to provide themeData.

You must edit it before returning a plugin api options.

**Reply by superewald** [reply]
Thanks for the reply but I don't understand what you refer with "You must edit it before returning a plugin api options".

Do you mean 
1. passing the navbar configuration as option to the @vuepress/plugin-theme-data plugin through config or 
2. directly changing the plugin itself ? 

With 1. I don't get access to frontmatter and page data from vuepress and I want to avoid using an additional frontmatter parser.
With 2. I don't understand how I can swap the native plugin with my own (despite some hacky solutions like removing the old plugin from config within the actual config files - very bad idea).

Honestly I don't even get to the point where configuration for the default-theme plugins is composed and thus have no clue how I could swap them. The only solution I currently can think of is to create a new theme (without extends) and configure all plugins on my own - which again would be tedious.

Maybe it would be enough for me to know where the @vuepress/plugin-theme-data gets its `themeData` option from as I failed to find that part.

*Edit*: I guess [defaultTheme.ts#30](https://github.com/vuepress/vuepress-next/blob/437b75076667e653d3600c96f9f4a7c3c3e47e57/packages/%40vuepress/theme-default/src/node/defaultTheme.ts#L30) `localeOptions` is what I want to change after `assignDefaultLocaleOptions` and before `['@vuepress/theme-data', { themeData: localeOptions }]` is executed but I don't see a way to grab `localeOptions` from within any plugin hook.

**Reply by Mister-Hope** [reply]
That's the problem. You are almost getting it. ThemeData is a temp file generated by another plugin in another word. So only your edits when initializing plugins(that is before returning a plugin API, has effects. Because that happens before the temp file is generated.

So if you are taking about reading pages data then modify theme config, you probably need to make some hack yourself. Such as replacing the temp file in your plugin.
